File: app/src/main/java/de/michael/filebinmobile/fragments/SettingsFragment.java
package de.michael.filebinmobile.fragments;

import android.app.Activity;
import android.app.AlertDialog;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.Nullable;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import de.michael.filebinmobile.R;
import de.michael.filebinmobile.adapters.OnAdapterDataChangedListener;
import de.michael.filebinmobile.adapters.ServerSettingsAdapter;
import de.michael.filebinmobile.controller.OnErrorOccurredCallback;
import de.michael.filebinmobile.controller.SettingsManager;
import de.michael.filebinmobile.controller.refactor.NetworkManager;
import de.michael.filebinmobile.model.PostInfo;
import de.michael.filebinmobile.model.Server;
import de.michael.filebinmobile.model.UserProfile;

public class SettingsFragment extends NavigationFragment implements OnAdapterDataChangedListener {

    @BindView(R.id.rclServerList)
    RecyclerView rclServerList;

    @BindView(R.id.txtEmptyServerList)
    TextView txtEmptyList;

    private ServerSettingsAdapter adapter;

    private CreateApikeyTask createApikeyTask;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        this.adapter = new ServerSettingsAdapter(getActivity());
        this.adapter.setDataChangedListener(this);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.server_settings_fragment, container, false);
        ButterKnife.bind(this, view);

        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getContext());
        linearLayoutManager.setOrientation(LinearLayoutManager.VERTICAL);

        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(this.rclServerList.getContext(),
                linearLayoutManager.getOrientation());

        this.rclServerList.setLayoutManager(linearLayoutManager);
        this.rclServerList.setItemAnimator(new DefaultItemAnimator());
        this.rclServerList.setAdapter(this.adapter);
        this.rclServerList.addItemDecoration(dividerItemDecoration);

        return view;
    }

    @Override
    public void onResume() {

        reloadServerList();

        super.onResume();
    }

    @Override
    public void onDestroy() {

        // cancel our async task in case it's running
        if (this.createApikeyTask != null) {
            this.createApikeyTask.cancel(true);
        }

        super.onDestroy();
    }

    @OnClick(R.id.fbaAddServer)
    void addServer() {
        final AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
        final View view = LayoutInflater.from(getContext())
                .inflate(R.layout.edit_server_settings, null);
        builder.setTitle("Add Server")
                .setView(view)
                .setPositiveButton(R.string.ok, (dialogInterface, i) -> {
                    EditText edtServerName = view.findViewById(R.id.edtEditName);
                    EditText edtServerAddr = view.findViewById(R.id.edtEditAddress);
                    EditText edtUserName = view.findViewById(R.id.edtUserName);
                    EditText edtUserPw = view.findViewById(R.id.edtUserPassword);

                    // create a new server model
                    Server server = new Server(
                            edtServerName.getText().toString(),
                            edtServerAddr.getText().toString()
                    );

                    // create a new user profile model
                    String userName = edtUserName.getText().toString();
                    String password = edtUserPw.getText().toString();

                    ApiKeyPostInfo apiKeyPostInfo = new ApiKeyPostInfo(userName, server, password);

                    createApikeyTask = new CreateApikeyTask(getActivity());
                    createApikeyTask.execute(apiKeyPostInfo);

                    Toast.makeText(getActivity(), "Creating Apikey", Toast.LENGTH_SHORT).show();

                })
                .setNegativeButton(R.string.cancel, (dialogInterface, i) -> dialogInterface.dismiss());
        AlertDialog dialog = builder.create();
        dialog.show();
    }

    private void reloadServerList() {

        // update active status of added server (even if user hasn't set the new server as
        // active, we still check here.
        PostInfo postInfo = SettingsManager.getInstance().getPostInfo(getActivity());
        this.adapter.setSelectedPostInfo(postInfo);

        ArrayList<Server> serverList = SettingsManager.getInstance().getServerList(getActivity());

        this.adapter.updateData(serverList);

        if (this.adapter.getItemCount() <= 0) {
            txtEmptyList.setVisibility(View.VISIBLE);
        } else {
            txtEmptyList.setVisibility(View.GONE);
        }


    }

    @Override
    public void onAdapterDataChanged() {
        reloadServerList();
    }

    @Override
    void cancelAllPossiblyRunningTasks() {
        if (this.createApikeyTask != null) {
            this.createApikeyTask.cancel(true);
        }
    }

    private class CreateApikeyTask extends AsyncTask<ApiKeyPostInfo, Integer, String> implements OnErrorOccurredCallback {

        private Activity activity;
        private ApiKeyPostInfo postInfo;

        public CreateApikeyTask(Activity activity) {
            this.activity = activity;
        }

        @Override
        protected String doInBackground(ApiKeyPostInfo... postInfos) {

            if (postInfos.length > 0) {

                this.postInfo = postInfos[0];

                String userName = this.postInfo.getUsername();
                String password = this.postInfo.getPassword();
                Server server = this.postInfo.getServer();


                return NetworkManager.getInstance()
                        .generateApiKey(userName, password, server.getAddr(), this);

            }

            return null;
        }

        @Override
        protected void onPostExecute(String apikey) {

            if (apikey != null) {

                Server server = this.postInfo.getServer();
                server.addUserProfile(new UserProfile(this.postInfo.username, apikey));

                SettingsManager.getInstance().addServer(server, getActivity());

                new AlertDialog.Builder(getContext())
                        .setTitle(R.string.serverAddedToList)
                        .setMessage(R.string.setServerAsActive)
                        .setPositiveButton(R.string.yes, ((dialogInterface, i) -> {
                            UserProfile userProfile = server.getUserProfiles().get(0);
                            PostInfo postInfo = new PostInfo(userProfile, server);
                            SettingsManager.getInstance()
                                    .setPostInfo(postInfo, getActivity());
                            reloadServerList();
                        }))
                        .setNegativeButton(R.string.no, ((dialogInterface, i) -> {
                            dialogInterface.dismiss();
                        }))
                        .create().show();

                reloadServerList();
            }

            super.onPostExecute(apikey);
        }

        @Override
        public void onError(String message) {
            // make sure we run this on the UI thread
            new Handler(Looper.getMainLooper()).post(() ->
                    Toast.makeText(this.activity, "Unable to create apikey", Toast.LENGTH_SHORT)
                            .show());
        }
    }

    /**
     * Wrapper class for information needed to create an apikey. Due to not saving the password in
     * the user model we can't use our normal PostInfo wrapper.
     */
    private class ApiKeyPostInfo {
        private Server server;
        private String username, password;

        public ApiKeyPostInfo(String username, Server server, String password) {
            this.username = username;
            this.server = server;
            this.password = password;
        }

        public String getUsername() {
            return username;
        }

        public Server getServer() {
            return server;
        }

        public String getPassword() {
            return password;
        }
    }
}

File: app/src/main/java/de/michael/filebinmobile/fragments/SettingsFragment.kt
package de.michael.filebinmobile.fragments

import android.app.AlertDialog
import android.content.DialogInterface
import android.graphics.PorterDuff
import android.os.AsyncTask
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Patterns
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import de.michael.filebinmobile.R
import de.michael.filebinmobile.adapters.ServerSettingsAdapter
import de.michael.filebinmobile.controller.NetworkManager
import de.michael.filebinmobile.controller.NetworkManager.updateServerInfo
import de.michael.filebinmobile.controller.SettingsManager
import de.michael.filebinmobile.model.Server
import de.michael.filebinmobile.model.UserProfile
import kotlinx.android.synthetic.main.edit_server_settings.view.*
import kotlinx.android.synthetic.main.server_settings_fragment.*
import kotlin.properties.Delegates

class SettingsFragment : NavigationFragment() {

    var adapter: ServerSettingsAdapter? = null

    private var createApiKeyTask: CreateApiKeyTask? by Delegates.observable(null) { _, oldVal: CreateApiKeyTask?, _: CreateApiKeyTask? ->
        // make sure we cancel the old task before creating a new one
        oldVal?.cancel(true)
    }

    private var loadServerInfoTask: LoadServerInfoTask? by Delegates.observable(null) { _, oldVal: LoadServerInfoTask?, _: LoadServerInfoTask? ->
        // make sure we cancel the old task before creating a new one
        oldVal?.cancel(true)
    }

    override fun cancelAllPossiblyRunningTasks() {
        createApiKeyTask = null
        loadServerInfoTask = null
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.server_settings_fragment, container, false)
                ?: super.onCreateView(inflater, container, savedInstanceState)
    }

    private fun init() {

        this.adapter = ServerSettingsAdapter(activity!!)

        rclServerList.setup()
        rclServerList.adapter = this.adapter

        fbaAddServer.setOnClickListener {
            val dialogView = LayoutInflater.from(activity)
                    .inflate(R.layout.edit_server_settings, null)

            var isValidUrl = false

            dialogView.edtEditAddress.addTextChangedListener(object : TextWatcher {
                override fun afterTextChanged(p0: Editable?) {
                    isValidUrl = Patterns.WEB_URL.matcher(p0?.toString()).matches()
                    if (isValidUrl) {
                        dialogView.edtEditAddress.background
                                .setColorFilter(resources.getColor(R.color.colorError), PorterDuff.Mode.SRC_IN)
                    } else {
                        dialogView.edtEditAddress.background
                                .setColorFilter(resources.getColor(R.color.colorAccent), PorterDuff.Mode.SRC_IN);
                    }
                }

                override fun beforeTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {
                    /* empty */
                }

                override fun onTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {
                    /* empty */
                }

            })
            AlertDialog.Builder(activity)
                    .setTitle("Add Server")
                    .setView(dialogView)
                    .setPositiveButton(R.string.ok) { _: DialogInterface, _: Int ->
                        if (isValidUrl) {
                            val serverName = dialogView.edtEditName.text.toString()
                            val serverAddr = dialogView.edtEditAddress.text.toString()
                            val userName = dialogView.edtUserName.text.toString()

                            val server = Server(serverAddr, serverName)

                            val apiKeyPostInfo = ApiKeyPostInfo(server, userName, dialogView.edtUserPassword.text.toString())

                            createApiKeyTask = CreateApiKeyTask()
                            createApiKeyTask?.execute(apiKeyPostInfo)

                            Toast.makeText(activity, "Creating Api Key", Toast.LENGTH_SHORT).show()
                        } else {
                            Toast.makeText(activity, "URL seems to be invalid", Toast.LENGTH_SHORT).show()
                        }
                    }
                    .setNegativeButton(R.string.cancel) { dialogInterface: DialogInterface, _: Int ->
                        dialogInterface.dismiss()
                    }
                    .create().show()
        }

    }

    override fun onStart() {
        init()
        reloadServerList()
        super.onStart()
    }

    override fun onResume() {
        reloadServerList()
        super.onResume()
    }

    private fun reloadServerList() {
        // update active status of added server (even if user hasn't set the new server as
        // active, we still check here)
        val postInfo = SettingsManager.getPostInfo(activity!!)
        this.adapter!!.selectedPostInfo = postInfo

        val serverList = SettingsManager.getServerList(activity!!)

        this.adapter!!.updateData(serverList)

        if (this.adapter!!.itemCount <= 0) {
            txtEmptyServerList.visibility = View.VISIBLE
        } else {
            txtEmptyServerList.visibility = View.GONE
        }
    }

    private inner class CreateApiKeyTask : AsyncTask<ApiKeyPostInfo, Int, String?>() {

        var postInfo: ApiKeyPostInfo? = null

        override fun doInBackground(vararg apiKeyPostInfo: ApiKeyPostInfo): String? {
            if (apiKeyPostInfo.isNotEmpty()) {

                this.postInfo = apiKeyPostInfo[0]

                return NetworkManager.generateApiKey(
                        apiKeyPostInfo[0].userName,
                        apiKeyPostInfo[0].password,
                        apiKeyPostInfo[0].server.address)
            }

            throw IllegalArgumentException("Not enough information to generate an api key")
        }

        override fun onPostExecute(result: String?) {
            if (!result.isNullOrBlank()) {

                val userProfile = UserProfile(this.postInfo!!.userName, result!!)
                val server = this.postInfo!!.server.copy(userProfile = userProfile)

                SettingsManager
                        .addServer(activity!!, server)

                AlertDialog.Builder(activity)
                        .setTitle(R.string.serverAddedToList)
                        .setMessage(R.string.setServerAsActive)
                        .setPositiveButton(R.string.yes) { _, i ->
                            SettingsManager.setPostInfo(activity!!, server)
                            reloadServerList()
                        }
                        .setNegativeButton(R.string.no) { dialogInterface, _ -> dialogInterface.dismiss() }
                        .create().show()

                // Load the server info and replace result from this task if success
                loadServerInfoTask = LoadServerInfoTask()
                loadServerInfoTask?.execute(server)

            }
            super.onPostExecute(result)
        }
    }

    /**
     * Due to the fact that this requires an extra api call we wrap it in it's own AsyncTask to
     * maintain some structure.
     * This results in extra work done by the settings manager, as we write to shared prefs after
     * adding the server and creating an api key and then overwrite the saved server after this task
     * finished successfully.
     */
    private inner class LoadServerInfoTask : AsyncTask<Server, Int, Pair<Server, Server>?>() {
        override fun doInBackground(vararg params: Server): Pair<Server, Server>? {
            if (params.isNotEmpty()) {
                val oldVal = params[0].copy()
                params[0].updateServerInfo()
                // we return a pair consisting of old value and new value so we can effectively
                // replace the current saved server
                return Pair(oldVal, params[0])
            }
            return null
        }

        override fun onPostExecute(result: Pair<Server, Server>?) {
            if (result != null) {
                SettingsManager.deleteServer(activity!!, result.first)
                SettingsManager.addServer(activity!!, result.second)
                val isCurrentUploadServer = result.first == SettingsManager.getPostInfo(activity!!)
                if (isCurrentUploadServer) {
                    // update our postInfo too
                    SettingsManager.setPostInfo(activity!!, result.second)
                }
            }
            super.onPostExecute(result)
        }
    }

}

private data class ApiKeyPostInfo(val server: Server, val userName: String, val password: String)